include/static-pages.php

```diff
@@ -122,23 +121,33 @@ public function page_link( $link, $id ) {
 	}
 
 	/**
-	 * Adds page_on_front and page_for_posts properties to the language objects.
+	 * Adds page_on_front and page_for_posts properties to language data before the object is created.
 	 *
-	 * @since 1.8
+	 * @since 3.4
+	 *
+	 * @param array $static_pages Array of language page_on_front and page_for_posts properties.
+	 * @param array $language Language data.
+	 * @return array {
+	 *     Static pages properties if exist.
 	 *
-	 * @param PLL_Language[] $languages The list of languages.
-	 * @param PLL_Model      $model     The instance of PLL_Model.
-	 * @return PLL_Language[]
+	 *     @type int $page_on_front  ID of the page on front in this language
+	 *     @type int $page_for_posts Id of the page for posts in this language
+	 * }
 	 */
-	public static function pll_languages_list( $languages, $model ) {
-		if ( 'page' === get_option( 'show_on_front' ) ) {
-			foreach ( $languages as $k => $language ) {
-				$languages[ $k ]->page_on_front = $model->post->get( get_option( 'page_on_front' ), $language );
-				$languages[ $k ]->page_for_posts = $model->post->get( get_option( 'page_for_posts' ), $language );
-			}
+	public function set_static_pages( $static_pages, $language ) {
+		$this->page_on_front  = intval( get_option( 'page_on_front' ) );
+		$this->page_for_posts = intval( get_option( 'page_for_posts' ) );
+
+		$page_on_front_translations  = $this->model->post->get_translations_from_term( $this->page_on_front );
+		$page_for_posts_translations = $this->model->post->get_translations_from_term( $this->page_for_posts );
+		if ( isset( $page_on_front_translations[ $language['slug'] ] ) ) {
+			$static_pages['page_on_front'] = $page_on_front_translations[ $language['slug'] ];
+		}
+		if ( isset( $page_for_posts_translations[ $language['slug'] ] ) ) {
+			$static_pages['page_for_posts'] = $page_for_posts_translations[ $language['slug'] ];
```

Comment:
You are doing 2 times the same thing with one different parameter. You can probably avoid duplicate code.

Reply:
Done in 4ae1f0b4410144953a2c550c14833fa9cea62b35